### ISSUE TITLE
sql: CREATE AS [SHOW RANGE FROM ... FOR ROW ...] job fails

### ISSUE THREAD

**Issue by ecwall** [cockroachdb/cockroach#106397]
**Setup** 
```
CREATE TABLE tbl (id int PRIMARY KEY);
```

**Statements causing failing jobs**
1. `CREATE TABLE t AS SELECT * FROM [SHOW RANGE FROM TABLE tbl FOR ROW (0)];`

The resulting error is:
```
ERROR: internal error: runtime error: invalid memory address or nil pointer dereference
SQLSTATE: XX000
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/util/errorutil/catch.go:29: ShouldCatch()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:196: func1()
GOROOT/src/runtime/panic.go:884: gopanic()
GOROOT/src/runtime/panic.go:260: panicmem()
GOROOT/src/runtime/signal_unix.go:835: sigpanic()
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/builtins.go:5069: func239()
github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:459: EvalFuncExpr()
github.com/cockroachdb/cockroach/bazel-out/darwin_arm64-fastbuild/bin/pkg/sql/sem/tree/eval_expr_generated.go:264: Eval()
github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:24: Expr()
github.com/cockroachdb/cockroach/pkg/sql/opt/norm/fold_constants_funcs.go:648: FoldFunction()
github.com/cockroachdb/cockroach/bazel-out/darwin_arm64-fastbuild/bin/pkg/sql/opt/norm/factory.og.go:19971: ConstructFunction()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/scalar.go:561: buildFunction()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/scalar.go:333: buildScalar()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/scalar.go:321: buildScalar()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/scalar.go:126: buildScalar()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/util.go:463: resolveAndBuildScalar()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:1169: buildWhere()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:1051: buildSelectClause()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:996: buildSelectStmtWithoutParens()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:965: func1()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/with.go:117: processWiths()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:964: buildSelect()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:316: buildStmt()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:411: buildStmt()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:168: buildDataSource()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:64: buildDataSource()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:1223: buildFromTablesRightDeep()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:1200: buildFromTables()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:1127: buildFrom()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:1048: buildSelectClause()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:996: buildSelectStmtWithoutParens()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:965: func1()
```

**Job Failure Error**
Error message is slightly different for `CREATE MATERIALIZED VIEW`, but job fails with 
```
SELECT job_id, status, error, description 
FROM [SHOW JOBS] 
WHERE job_type IN ('SCHEMA CHANGE', 'NEW SCHEMA CHANGE')
AND status != 'succeeded';

880373734131400705, failed, runtime error: invalid memory address or nil pointer dereference, CREATE TABLE defaultdb.public.t (start_key STRING, end_key STRING, range_id INT8, lease_holder INT8, lease_holder_locality STRING, replicas INT8[], replica_localities STRING[], voting_replicas INT8[], non_voting_replicas INT8[]) AS SELECT * FROM [SHOW RANGE FROM TABLE defaultdb.public.tbl FOR ROW (0)]
```

Jira issue: CRDB-29554

**Comment by rafiss** [comment]
The root cause is that `evalCtx.CatalogBuiltins` is nil when the `crdb_internal.encode_key` builtin function is executed by the optimizer. That makes this line fail: https://github.com/cockroachdb/cockroach/blob/a6dbdee0e07325f58a8baf30c199c9cf7ad46614/pkg/sql/sem/builtins/builtins.go#L5126

**Comment by mgartner** [comment]
So the `CREATE TABLE AS` creates a job, and a query run by the job fails? Is that correct?

**Comment by yuzefovich** [comment]
I fixed it independently in #107279, will pull out a separate commit.
